src/tests/adr/initializer_functions_adr.rs

```diff
@@ -0,0 +1,432 @@
+use insta::assert_debug_snapshot;
+use plc_ast::provider::IdProvider;
+
+use crate::{
+    index::const_expressions::UnresolvableKind,
+    resolver::const_evaluator::evaluate_constants,
+    test_utils::tests::{annotate_and_lower_with_ids, index, index_with_ids},
+};
+
+/// # Architecture Design Records: Lowering of complex initializers to initializer functions
+///
+/// When encountering an unresolvable initializer to a pointer during constant propagation,
+/// rusty will mark this const-expression for a retry during later stages in the compilation pipeline.
```

Comment:
While reviewing this PR I was thinking to myself if the implementation of this would have been easier if we lowered / lazily initialized POU members instead of tracking and introducing `__init_...` functions via the index and resolver. For example every POU would get a `__initialized` boolean flag and a POU such as 
```
FUNCTION_BLOCK foo
  VAR
    mem : REF_TO DINT := REF(...)
  END_VAR
END_FUNCTION_BLOCK
```
gets lowered to 
```
FUNCTION_BLOCK foo
  VAR
    __initialized : BOOL;
    mem : REF_TO DINT;
  END_VAR

  IF NOT __initialized THEN
    mem := REF(...)
    __initialized := TRUE;
  END_IF
```

Obviously we don't have the lowering infrastructure yet and this would not work for structs with default value fields but just thinking out loud here

Reply:
One of the reasons for not doing initialization in the POU bodies is the performance hit during first cycle. The major advantage of the current implementation, with all initializers being standalone POUs, is that the runtime can call and initialize all POUs before starting main task execution, eliminating initialization as a factor of task execution time variance.

Reply:
Surely assignments with REF function calls won't have much of a performance impact or am I missing something?

Reply:
it might not be much but it affects the real-time behavior. At this point it's not performance that matters but runtime consistency.